File: Character.py
from typing import List

class Character:
    """
    It represents a character.

    Contains all the infos to be stored about a character.
    """
    _id = 0

    #Inizializza un'istanza dell'oggetto
    def __init__(self):
        self._max_range = 0
        self._factions = []
        self._health = 0
        self._level = 0
        self._isAlive = False
        self._damage = 0
        Character._id += 1
        self._uid = Character._id

    def get_damage(self):
        return self._damage
    
    def get_health(self):
        return self._health
    
    def get_level(self):
        return self._level

    def set_max_range(self, range: int) -> None:
        self._max_range = range
    
    def set_health(self, health: int) -> None:
        self._health = health
    
    def set_alive(self, alive: bool) -> None:
        self._isAlive = alive
    
    def set_level(self, level: int) -> None:
        self._level = level
    
    def add_faction(self, faction) -> None:
        self._factions.append(faction)
    
    def get_factions(self) -> List[int]:
        return self._factions
    
    def get_range_max(self) -> int:
        return self._max_range

    def __del__(self):
        self._factions.clear()

    def __eq__(self, other) -> bool:
        if isinstance(other, Character):
            return self._uid == other._uid
        return False

File: cpp/character.cpp
#include "character.hpp"
#include "exceptions.hpp"
#include "constants.hpp"
#include "utils.hpp"
#include <iostream>

int Character::_id = 0;

Character::Character(const Position& p) : _health(1000), _level(1), _isAlive(true) {
    ++_id;
    _pos = new Position(); // never forget to allocate when using pointers!
    _pos->x = p.x;
    _pos->y = p.y;
}

Character::Character() {}

Character::~Character() {
    delete _pos;
    --_id;
}

// copy constructor
Character::Character(const Character& character) : _health(character.getHealth()), _level(character.getLevel()), _isAlive(character.isAlive()) {
    _pos = new Position();
    *_pos = character.getPosition();
}

void Character::setAlive(bool alive) { _isAlive = alive; }

void Character::setHealth(int health) { _health = health; }

void Character::setLevel(int level) { _level = level; }

void Character::addFaction(int faction) { _factions.push_back(faction); }

int Character::getHealth() const { return _health; }

bool Character::isAlive() const { return _isAlive; }

int Character::getLevel() const { return _level; }

std::vector<int> Character::getFactions() const { return _factions; }

std::string Character::getDescription() const { return "A character "; }

void Character::setMaxRange(int range) { _max_range = range; }

int Character::getRangeMax() const { return _max_range; }

Position& Character::getPosition() const { return *_pos; }

bool Character::operator==(const Character& character) { return character._id == this->_id; }